java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -170,7 +170,10 @@ class DatabaseImpl(
 
     override suspend fun removeClient(identifier: Int) = clientMutex.withLock {
         clients.remove(nextClientId)
-        // TODO: When all clients are done with the database, close the connection.
+        // When all clients are done with the database, close the connection.
+        if (clients.isEmpty()) {
+            super.close()
```

Comment:
this causes problems, for bulk deletes we use this class but we don't register clients. If the last client goes away when we are doing some bulk delete, we get an exception. Why is this necessary?

Reply:
If you never invoke removeClient() because you never invoked addClient(), why is it a problem?

I'll defer to ian, but probably there's some benefit to not keeping SQLite open.

A new onDatabaseClose() lambda callback has been added since this PR, so the 'super.close()' could be moved into the lambda, and if you are a client which doesn't use add/removeClients(), you can just not supply an onDatabaseClose which closes the DB. 

Although if the DB is being used without the add/remove client stuff, maybe the DatabaseImpl needs to be refactored in a type for global DB management stuff (e.g. bulk) vs one that's a client for the storage layer.

Reply:
It this what happens is:
1)client A calls addClient
2)bulk delete starts
3)client A calls removeClient. since there are not clients, db gets closed. But the bulk delete has not finished.

so even if you move the super.close() up one class you still have problems.
bulk delete uses a database but not a database driver (which manages registrations).
Maybe we can register a fake client with a made up storageKey?

Reply:
or maybe using a separate instance is better

Reply:
You could do that, or I'd support removing the SQLite.close() call (it would still have to invoke onDatabaseClose() though, because that cleans up actual memory issues), provided I run a heap measurement to make sure it doesn't introduce a leak. (a bounded increase I guess is temporarily ok). I don't want you to be blocked some minor memory consumption (e.g. lets say it adds O(1 mb) overhead)

However, @ianull should chime in if there's any other gotchas.

Reply:
sqlite3 does caching `per connection` (i'm not sure the default settings of size/pages for sqlite@Android).

The downside without closing a connection is the cache will not be cleared until the process is killed, though I remembered there is an api to ask sqlite for clearing up caches as much as possible)

I like the idea of using a separate instance which establish another connection to database. It would be neat without worrying about race condition.
Open a connection, do the bulk writes then close the connection.

Reply:
Ok I am going to try this thanks.